File: app/src/main/java/com/example/exemplocardview/componenteCardView/CardQuestion.java
package com.example.exemplocardview.componenteCardView;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import androidx.constraintlayout.widget.ConstraintLayout;
import com.example.exemplocardview.R;

import org.jetbrains.annotations.Nullable;

public class CardQuestion extends LinearLayout implements CardQuestionInterface {

    public static final int DURATION = 300;
    public static final int ROTATION_90 = 90;
    public static final int ROTATION_270 = 270;
    private boolean isCardOpen = true;
    private boolean isLikeIconSet = false;
    private boolean isUnlikeIconSet = false;

    private ConstraintLayout mLayoutHeader, mLayoutBodyBlock3;
    private LinearLayout mLayoutBodyBlocks, mLayoutBody, mLayoutBodyBlock1, mLayoutBodyBlock2;
    private ImageButton mUnlikeButton, mLikeButton;
    private ImageView mIconImageHeader, mArrowImageheader;
    private TextView mTextQuestion, mCounterTextRemarks, mTextHeader;
    private EditText mTextRemarks;

    private OnClickHeaderListener onHeaderClickListener;
    private OnClickLikeButtonListener onLikeButtonClickListener;
    private OnClickUnlikeButtonListener onUnlikeButtonClickListener;

    private String mTextBlock1, mValueBlock1, mTextBlock2, mValueBlock2, mTextBlock3, mValueBlock3, mTitleHeader;

    public CardQuestion(Context context) {
        super(context);
        this.init(context);
    }

    public CardQuestion(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.init(context);
    }

    public CardQuestion(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.init(context);
    }

    public void onClickHeader(OnClickHeaderListener onHeaderClickListener) {
        this.onHeaderClickListener = onHeaderClickListener;
    }

    public void onUnlikeClickButton(OnClickUnlikeButtonListener onUnlikeButtonClickListener) {
        this.onUnlikeButtonClickListener = onUnlikeButtonClickListener;
    }

    public void onLikeClickButton(OnClickLikeButtonListener onLikeButtonClickListener) {
        this.onLikeButtonClickListener = onLikeButtonClickListener;
    }

    public String getTitleHeader() {
        return mTitleHeader;
    }

    public void setTitleHeader(String mTitleHeader) {
        this.mTitleHeader = mTitleHeader;
    }

    public interface OnClickHeaderListener {
        void onClick(boolean bool);
    }

    public interface OnClickLikeButtonListener {
        void onClick();
    }
    public interface OnClickUnlikeButtonListener {
        void onClick();
    }

    private void setValueBlock1(String mValueBlock1) {
        this.mValueBlock1 = mValueBlock1;
    }

    private void setValueBlock2(String mValueBlock2){
        this.mValueBlock2 = mValueBlock2;
    }

    private void setValueBlock3(String mValueBlock3){
        this.mValueBlock3 = mValueBlock3;
    }

    private void init(Context context) {
        View cardRoot = View.inflate(context, R.layout.card_question, this);

        //Header
        mLayoutHeader = cardRoot.findViewById(R.id.layout_header);
        mIconImageHeader = cardRoot.findViewById(R.id.icon_image_header);
        mArrowImageheader = cardRoot.findViewById(R.id.arrow_image_header);
        mTextHeader = cardRoot.findViewById(R.id.text_header);

        //Body
        mLayoutBody = cardRoot.findViewById(R.id.layout_body);
        mLayoutBodyBlocks = cardRoot.findViewById(R.id.layout_body_blocks);

        mLayoutBodyBlock1 = cardRoot.findViewById(R.id.layout_body_block1);
        mTextQuestion = cardRoot.findViewById(R.id.text_question);

        mLayoutBodyBlock2 = cardRoot.findViewById(R.id.layout_body_block2);
        mUnlikeButton = cardRoot.findViewById(R.id.unlike_button);
        mLikeButton = cardRoot.findViewById(R.id.like_button);

        mLayoutBodyBlock3 = cardRoot.findViewById(R.id.layout_body_block3);
        mTextRemarks = cardRoot.findViewById(R.id.text_remarks);
        mCounterTextRemarks = cardRoot.findViewById(R.id.text_counter);

        setClicks();

        updateTextCounter();
    }

    private void setClicks() {
        mLayoutHeader.setOnClickListener(getOnClickLayoutHeader());
        mLikeButton.setOnClickListener(getOnClickLikeButton());
        mUnlikeButton.setOnClickListener(getOnClickUnlikeButton());
    }

    private OnClickListener getOnClickLikeButton() {
        return new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (onLikeButtonClickListener != null) {
                    onLikeButtonClickListener.onClick();
                }
            }
        };
    }

    private OnClickListener getOnClickUnlikeButton() {
        return new OnClickListener() {
            @Override
            public void onClick(View v) {
                if(onUnlikeButtonClickListener != null){
                    onUnlikeButtonClickListener.onClick();
                }
            }
        };
    }

    private OnClickListener getOnClickLayoutHeader() {
        return new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (onHeaderClickListener != null) {
                    onHeaderClickListener.onClick(isCardOpen());
                }
            }
        };
    }

    private void openAnim() {
        ObjectAnimator.ofFloat(mArrowImageheader, View.ROTATION, ROTATION_270)
                .setDuration(DURATION).start();
    }

    private void closeAnim() {
        ObjectAnimator.ofFloat(mArrowImageheader, View.ROTATION, ROTATION_90)
                .setDuration(DURATION).start();
    }

    private void setValueOnBlock1(String value) {
        if (TextUtils.isEmpty(value)) {
            mTextQuestion.setVisibility(GONE);
        } else {
            setValueBlock1(value);
            mTextQuestion.setText(value);
        }
    }

    private void setValueOnBlock2(String value) {
        if (TextUtils.isEmpty(value)) {
            //TODO - Giovani - Disable visibility on components used on this block2
        } else {
            setValueBlock2(value);
            //TODO - Giovani - Set values on components used on this block2
        }
    }

    private void setValueOnBlock3(String value){
        if(TextUtils.isEmpty(value)){
            mTextRemarks.setVisibility(GONE);
        }else{
            setValueBlock3(value);
            mTextRemarks.setText(value);
        }
    }

    private void updateTextCounter(){
        mTextRemarks.addTextChangedListener(new CharacterWatcher() {
            @Override
            public void afterCharacterChanged(@Nullable Character character, @Nullable Integer count) {
                mCounterTextRemarks.setText(count.toString() + "/130");
            }
        });
    }

    @Override
    public void openCard() {
        isCardOpen = true;
        mLayoutHeader.setContentDescription("Recolher " + getTitleHeader());
        CardQuestionAnimation.expand(mLayoutBody);
        openAnim();
    }

    @Override
    public void closeCard() {
        isCardOpen = false;
        mLayoutHeader.setContentDescription("Expandir " + getTitleHeader());
        CardQuestionAnimation.collapse(mLayoutBody);
        closeAnim();
    }

    @Override
    public boolean isCardOpen() {
        return isCardOpen;
    }

    @Override
    public boolean isCardClose() {
        return isCardOpen;
    }

    @Override
    public void textHeader(@Nullable String text) {
        if (TextUtils.isEmpty(text)) {
            mTextHeader.setVisibility(GONE);
        } else {
            setTitleHeader(text);
            mTextHeader.setText(getTitleHeader());
        }
    }

    @Override
    public void textQuestion(@Nullable String text) {
        if (TextUtils.isEmpty(text)) {
            mLayoutBodyBlock1.setVisibility(GONE);
        } else {
            mLayoutBodyBlock1.setVisibility(VISIBLE);
            setValueOnBlock1(text);
        }
    }

    @Override
    public void textRemarks(@Nullable String text) {
        if (TextUtils.isEmpty(text)) {
            mLayoutBodyBlock3.setVisibility(GONE);
        } else {
            mLayoutBodyBlock3.setVisibility(VISIBLE);
            setValueOnBlock3(text);
        }
    }

    @Override
    public void setLikeIconActivated() {
        isLikeIconSet = true;
        mLikeButton.setImageResource(R.drawable.ic_thumbs_up_icon_2);
        mIconImageHeader.setImageResource(R.drawable.ic_thumbs_up_icon_2);
    }

    @Override
    public void setLikeIconDeactivated() {
        isLikeIconSet = false;
        mLikeButton.setImageResource(R.drawable.ic_thumbs_up_icon);
    }

    @Override
    public void setUnlikeIconActivated() {
        isUnlikeIconSet = true;
        mUnlikeButton.setImageResource(R.drawable.ic_thumbs_down_icon_2);
        mIconImageHeader.setImageResource(R.drawable.ic_thumbs_down_icon_2);
    }

    @Override
    public void setUnlikeIconDeactivated() {
        isUnlikeIconSet = false;
        mUnlikeButton.setImageResource(R.drawable.ic_thumbs_down_icon);
    }

    @Override
    public boolean isLikeSet() {
        return isLikeIconSet;
    }

    @Override
    public boolean isUnlikeSet() {
        return isUnlikeIconSet;
    }

    @Override
    public void resetComponent() {
        closeCard();
        mIconImageHeader.setImageResource(R.drawable.ic_thumbs_up_icon);
        mLikeButton.setImageResource(R.drawable.ic_thumbs_up_icon);
        mUnlikeButton.setImageResource(R.drawable.ic_thumbs_down_icon);
        mTextHeader.setText("");
        mTextQuestion.setText("");
        mTextRemarks.setText("");
    }
}

File: app/src/main/java/com/example/exemplocardview/componenteCardView/CardQuestion.kt
package com.example.exemplocardview.componenteCardView

import android.animation.ObjectAnimator
import android.content.Context
import android.content.res.Resources
import android.os.Parcel
import android.os.Parcelable
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.AttributeSet
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import androidx.constraintlayout.widget.ConstraintLayout
import com.example.exemplocardview.R

class CardQuestion : LinearLayout {

    private var mLayoutHeader: ConstraintLayout? = null
    private var mIconImageHeader: ImageView? = null
    private var mArrowImageheader: ImageView? = null
    private var mLayoutBody: LinearLayout? = null
    private var mLikeButton: ImageButton? = null
    private var mUnlikeButton: ImageButton? = null
    private var mTextHeader: TextView? = null
    private var mTextQuestion: TextView? = null
    private var mCounterTextRemarks: TextView? = null
    private var mTextRemarks: EditText? = null

    var isCardOpen: Boolean = true
        set(value) {
            field = value
            openCloseCard(field)
        }

    var status: Int = 0
        set(value) {
            field = value
            configStatus(field)
        }

    var header: String = ""
        set(value) {
            field = value
            mTextHeader?.text = field
        }
    var question: String = ""
        set(value) {
            field = value
            mTextQuestion?.text = field

            if (TextUtils.isEmpty(field.trim())) {
                mTextQuestion?.visibility = View.GONE
            } else {
                mTextQuestion?.visibility = View.VISIBLE
            }
        }

    var remarks: String = ""
        set(value) {
            field = value

            mTextRemarks?.let {
                it.removeTextChangedListener(onRemarksChange)
                it.setText(field)
                it.setSelection(field.length)
                it.addTextChangedListener(onRemarksChange)
                updateCounter(field)
            }
        }

    private fun configStatus(status: Int) {
        when (status) {
            0 -> {
                mIconImageHeader?.setImageResource(R.drawable.ic_thumbs_up_icon)
                mLikeButton?.setImageResource(R.drawable.ic_thumbs_up_icon)
                mUnlikeButton?.setImageResource(R.drawable.ic_thumbs_down_icon)
            }
            1 -> {
                mLikeButton?.setImageResource(R.drawable.ic_thumbs_up_icon_2)
                mUnlikeButton?.setImageResource(R.drawable.ic_thumbs_down_icon)
                mIconImageHeader?.setImageResource(R.drawable.ic_thumbs_up_icon_2)
            }
            2 -> {
                mLikeButton?.setImageResource(R.drawable.ic_thumbs_up_icon)
                mUnlikeButton?.setImageResource(R.drawable.ic_thumbs_down_icon_2)
                mIconImageHeader?.setImageResource(R.drawable.ic_thumbs_down_icon_2)
            }
        }
    }

    private fun openCloseCard(status: Boolean) {
        mLayoutBody?.animateVisibility(status)

        if (status) {
            openAnim()
        } else {
            closeAnim()
        }
    }

    private fun openAnim() {
        ObjectAnimator.ofFloat(
            mArrowImageheader,
            View.ROTATION,
            ROTATION_270.toFloat()
        )
            .setDuration(DURATION.toLong()).start()
    }

    private fun closeAnim() {
        ObjectAnimator.ofFloat(
            mArrowImageheader,
            View.ROTATION,
            ROTATION_90.toFloat()
        )
            .setDuration(DURATION.toLong()).start()
    }

    private fun updateCounter(text: String) {
        val res: Resources = resources
        var counter= res.getString(R.string.card_remarks_counter_max, text.length.toString())
        mCounterTextRemarks?.text = counter
    }

    constructor(context: Context?) : super(context) {
        initialize(null)
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        initialize(attrs)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        initialize(attrs)
    }

    private fun initialize(attrs: AttributeSet?) {

        val inflater = context
            .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val cardRoot = inflater.inflate(R.layout.card_question, this)

        //Header
        mLayoutHeader = cardRoot.findViewById(R.id.layout_header)
        mIconImageHeader = cardRoot.findViewById(R.id.icon_image_header)
        mArrowImageheader = cardRoot.findViewById(R.id.arrow_image_header)

        //Body
        mLayoutBody = cardRoot.findViewById(R.id.layout_body)


        mTextHeader = cardRoot.findViewById(R.id.text_header)
        mTextQuestion = cardRoot.findViewById(R.id.text_question)
        mTextRemarks = cardRoot.findViewById(R.id.text_remarks)
        mCounterTextRemarks = cardRoot.findViewById(R.id.text_counter)

        //Buttons
        mUnlikeButton = cardRoot.findViewById(R.id.unlike_button)
        mLikeButton = cardRoot.findViewById(R.id.like_button)

        setClicks()

        attrs?.let {
            val typedArray =
                context.obtainStyledAttributes(it, R.styleable.CardQuestion, 0, 0)

            isCardOpen = typedArray.getBoolean(R.styleable.CardQuestion_isCardOpen, true)
            status = typedArray.getInt(R.styleable.CardQuestion_status, 0)
            header = typedArray.getString(R.styleable.CardQuestion_header) ?: ""
            question = typedArray.getString(R.styleable.CardQuestion_question) ?: ""
            remarks = typedArray.getString(R.styleable.CardQuestion_remarks) ?: ""
        }
    }

    private fun setClicks() {
        mLayoutHeader?.setOnClickListener {
            isCardOpen = !isCardOpen
        }

        mLikeButton?.setOnClickListener {
            status = 1
        }

        mUnlikeButton?.setOnClickListener {
            status = 2
        }

        mTextRemarks?.addTextChangedListener(onRemarksChange)
    }

    private val onRemarksChange = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {
            remarks = s.toString()
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            // Sem Uso
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            // Sem Uso
        }

    }

    fun resetComponent() {
        isCardOpen = false
        status = 0
        header = ""
        question = ""
        remarks = ""
    }

    companion object {
        const val DURATION = 5
        const val ROTATION_90 = 90
        const val ROTATION_270 = 270
    }

    override fun dispatchSaveInstanceState(container: SparseArray<Parcelable>) {
        super.dispatchFreezeSelfOnly(container)
    }

    override fun dispatchRestoreInstanceState(
        container: SparseArray<Parcelable>
    ) {
        super.dispatchThawSelfOnly(container)
    }

    override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState()
        val ss = SavedState(superState!!)
        ss.isCardOpen = isCardOpen
        ss.status = status
        ss.header = header
        ss.question = question
        ss.remarks = remarks
        return ss
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        val ss = state as SavedState
        super.onRestoreInstanceState(ss.superState)
        //
        isCardOpen = ss.isCardOpen
        status = ss.status
        header = ss.header
        question = ss.question
        remarks = ss.remarks
    }

    /**
     * Saved State inner static class
     */
    private class SavedState : View.BaseSavedState {
        var isCardOpen: Boolean = true
        var status: Int = 0
        var header: String = ""
        var question: String = ""
        var remarks: String = ""

        constructor(superState: Parcelable) : super(superState)

        private constructor(source: Parcel) : super(source) {
            isCardOpen = source.readInt() != 0
            status = source.readInt()
            header = source.readString()!!
            question = source.readString()!!
            remarks = source.readString()!!
        }

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)

            out.writeInt(if (isCardOpen) 1 else 0)
            out.writeInt(status)
            out.writeString(header)
            out.writeString(question)
            out.writeString(remarks)
        }

        companion object CREATOR : Parcelable.Creator<SavedState> {

            override fun createFromParcel(parcel: Parcel): SavedState {
                return SavedState(parcel)
            }

            override fun newArray(size: Int): Array<SavedState?> {
                return arrayOfNulls(size)
            }

        }

        override fun describeContents(): Int {
            return 0
        }
    }
}